### PR TITLE
Update Electron - MSAL Node sample to use latest version of Electron

### LINKED ISSUE #4631: msal node test app and electron 17
### Core Library

MSAL Node (@azure/msal-node)

### Core Library Version

1.70

### Wrapper Library

MSAL Node Extensions (@azure/msal-node-extensions)

### Wrapper Library Version

1.70

### Description

When taking the electron sample and updating electron to 17.2.0 I am getting the follow error when logging in and trying to open emails.

### Error Message

E:\MICROSOFT_WORK\microsoft-authentication-library-for-js\samples\msal-node-samples\ElectronTestApp>npm start

> electron-test-app@1.0.0 start E:\MICROSOFT_WORK\microsoft-authentication-library-for-js\samples\msal-node-samples\ElectronTestApp
> npm run build && electron dist/app.js

> electron-test-app@1.0.0 build E:\MICROSOFT_WORK\microsoft-authentication-library-for-js\samples\msal-node-samples\ElectronTestApp
> tsc

[Fri, 25 Mar 2022 21:34:43 GMT] : @azure/msal-node@1.7.0 : Info - getTokenCache called
[Fri, 25 Mar 2022 21:34:48 GMT] : @azure/msal-node@1.7.0 : Info - getAuthCodeUrl called
[Fri, 25 Mar 2022 21:34:48 GMT] : @azure/msal-node@1.7.0 : Verbose - initializeRequestScopes called
[Fri, 25 Mar 2022 21:34:48 GMT] : [84fb7100-2d07-4ac1-a864-1f1d620e11a3] : @azure/msal-node@1.7.0 : Verbose - buildOauthClientConfiguration called
[Fri, 25 Mar 2022 21:34:48 GMT] : [84fb7100-2d07-4ac1-a864-1f1d620e11a3] : @azure/msal-node@1.7.0 : Verbose - building oauth client configuration with the authority: https://login.microsoftonline.com/08416161-211e-4d03-a6f5-4e12360d74c3
[Fri, 25 Mar 2022 21:34:48 GMT] : [84fb7100-2d07-4ac1-a864-1f1d620e11a3] : @azure/msal-node@1.7.0 : Verbose - createAuthority called
[Fri, 25 Mar 2022 21:34:49 GMT] : [84fb7100-2d07-4ac1-a864-1f1d620e11a3] : @azure/msal-node@1.7.0 : Verbose - Auth code client created
[Fri, 25 Mar 2022 21:34:50 GMT] : @azure/msal-node@1.7.0 : Info - acquireTokenByCode called
[Fri, 25 Mar 2022 21:34:50 GMT] : @azure/msal-node@1.7.0 : Verbose - initializeRequestScopes called
[Fri, 25 Mar 2022 21:34:50 GMT] : [7365176d-273a-4c60-882b-11cee9b4d959] : @azure/msal-node@1.7.0 : Verbose - buildOauthClientConfiguration called
[Fri, 25 Mar 2022 21:34:50 GMT] : [7365176d-273a-4c60-882b-11cee9b4d959] : @azure/msal-node@1.7.0 : Verbose - building oauth client configuration with the authority: https://login.microsoftonline.com/08416161-211e-4d03-a6f5-4e12360d74c3
[Fri, 25 Mar 2022 21:34:50 GMT] : [7365176d-273a-4c60-882b-11cee9b4d959] : @azure/msal-node@1.7.0 : Verbose - createAuthority called
[Fri, 25 Mar 2022 21:34:50 GMT] : [7365176d-273a-4c60-882b-11cee9b4d959] : @azure/msal-node@1.7.0 : Verbose - Auth code client created
[Fri, 25 Mar 2022 21:34:50 GMT] : [7365176d-273a-4c60-882b-11cee9b4d959] : @azure/msal-common@6.1.0 : Info - in acquireToken call
(node:163884) electron: Failed to load URL: msal://redirect/?code=<TOKEN_REDACTED>&session_state=165060c1-98e2-488d-b31a-40d0501c4e72 with error: ERR_FILE_NOT_FOUND
(Use `electron --trace-warnings ...` to show where the warning was created)
[Fri, 25 Mar 2022 21:34:50 GMT] : [7365176d-273a-4c60-882b-11cee9b4d959] : @azure/msal-common@6.1.0 : Verbose - Persistence enabled, calling beforeCacheAccess
[Fri, 25 Mar 2022 21:34:50 GMT] : [7365176d-273a-4c60-882b-11cee9b4d959] : @azure/msal-common@6.1.0 : Verbose - Persistence enabled, calling afterCacheAccess
[Fri, 25 Mar 2022 21:34:55 GMT] : @azure/msal-node@1.7.0 : Verbose - initializeRequestScopes called
[Fri, 25 Mar 2022 21:34:55 GMT] : [c0d494d7-a0f5-4d8e-944b-4bd117d97410] : @azure/msal-node@1.7.0 : Verbose - buildOauthClientConfiguration called
[Fri, 25 Mar 2022 21:34:55 GMT] : [c0d494d7-a0f5-4d8e-944b-4bd117d97410] : @azure/msal-node@1.7.0 : Verbose - building oauth client configuration with the authority: https://login.microsoftonline.com/08416161-211e-4d03-a6f5-4e12360d74c3
[Fri, 25 Mar 2022 21:34:55 GMT] : [c0d494d7-a0f5-4d8e-944b-4bd117d97410] : @azure/msal-node@1.7.0 : Verbose - createAuthority called
[Fri, 25 Mar 2022 21:34:55 GMT] : [c0d494d7-a0f5-4d8e-944b-4bd117d97410] : @azure/msal-node@1.7.0 : Verbose - Silent flow client created
Request made at: Fri Mar 25 2022 14:34:55 GMT-0700 (Pacific Daylight Time)
(node:163884) UnhandledPromiseRejectionWarning: Error: Request failed with status code 401
    at createError (E:\MICROSOFT_WORK\microsoft-authentication-library-for-js\samples\msal-node-samples\ElectronTestApp\node_modules\axios\lib\core\createError.js:16:15)
    at settle (E:\MICROSOFT_WORK\microsoft-authentication-library-for-js\samples\msal-node-samples\ElectronTestApp\node_modules\axios\lib\core\settle.js:17:12)
    at IncomingMessage.handleStreamEnd (E:\MICROSOFT_WORK\microsoft-authentication-library-for-js\samples\msal-node-samples\ElectronTestApp\node_modules\axios\lib\adapters\http.js:269:11)
    at IncomingMessage.emit (node:events:402:35)
    at endReadableNT (node:internal/streams/readable:1343:12)
    at processTicksAndRejections (node:internal/process/task_queues:83:21)
(node:163884) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)

### Msal Logs

_No response_

### MSAL Configuration

```javascript
{
    "authOptions":
        {
            "clientId": "904df931-49c6-404f-ad6c-4fd84a25d1bb",
            "authority": "https://login.microsoftonline.com/08416161-211e-4d03-a6f5-4e12360d74c3"
        },
    "request":
    {
        "authCodeUrlParameters": {
            "scopes": ["user.read"],
            "redirectUri": "msal://redirect"
        },
        "authCodeRequest": {
            "redirectUri": "msal://redirect",
            "scopes": ["User.Read"]
        }
    },
    "resourceApi":
    {
        "endpoint": "https://graph.microsoft-ppe.com/v1.0/me"
    },
    "fileProtocol":
    {
        "name": "msal"
    }
}
```

### Relevant Code Snippets

```javascript
updating the test msal node app to use electron 17.2.0. only two files were touched.

in createMainWindow in Main.ts I added
contextIsolation: false to the webpreferences.

in package.json I changed electron to
    "electron": "^17.2.0",
```

### Reproduction Steps

Clone the msal repo.

change electron to 17.2.0 in package.json

change contextIsolation to false in CreateMainWindow from the file Main.ts(electron set this default to true after v10)

npm start

login

attempt to open email

### Expected Behavior

error is not thrown and access is allowed.

### Identity Provider

Azure AD / MSA

### Browsers Affected (Select all that apply)

None (Server)

### Regression

_No response_

### Source

Internal (Microsoft)

### PR DESCRIPTION
**Pull request by hectormmg**
This PR:

- Addresses #4631
- Updates the communication method between Main and Renderer process to use the Electron `contextBridge` in the `preload.ts` script to give the Renderer process controlled and secured access to the Node APIs the Main process has access to
- Delegates authentication window creation and management to the `AuthProvider` class
- Refactors `CustomFileProtocolListener` into `CustomProtocolListener` since the internals now register a string protocol instead of a file protocol to listen for the auth code and avoid the protocol sending a file and causing a navigation error as a side effect of a successful authorization request.

### CODE REVIEW

File: samples/msal-node-samples/ElectronTestApp/src/AuthProvider.ts [hunk 1]
```diff
@@ -57,6 +57,14 @@ export default class AuthProvider {
         return this.account;
     }
 
+    // Creates main application window
```
Comment (hectormmg) [inline samples/msal-node-samples/ElectronTestApp/src/AuthProvider.ts]:
```suggestion
    // Creates a  "popup" window for interactive authentication
```

**Review by salman90 — APPROVED** [review]
LGTM!

**Review by jasonnutter — APPROVED** [review]

**Review by tnorling — APPROVED** [review]

**Review by sameerag — APPROVED** [review]
Looks good. q: Do we know if this is backward compatible?

**Comment by hectormmg** [comment]
@sameerag it does seem to be backward compatible from my own testing.

**Pull request merged**